### PR TITLE
Auth with OS error

### PR DESCRIPTION
**Pull request by stackchain**
- Auth OS errors will always be displayed using the Alert component
- RN keychain doesn't normalize iOS/Android
- Fix the android build

### CODE REVIEW

File: android/build.gradle [hunk 1]
```diff
@@ -33,7 +33,16 @@ buildscript {
     }
 }
 
+def REACT_NATIVE_VERSION = new File(['node', '--print',"JSON.parse(require('fs').readFileSync(require.resolve('react-native/package.json'), 'utf-8')).version"].execute(null, rootDir).text.trim())
+
 allprojects {
+    // https://github.com/facebook/react-native/issues/35210
+    configurations.all {
+        resolutionStrategy {
+            force "com.facebook.react:react-native:" + REACT_NATIVE_VERSION
```
Comment (stackchain) [inline android/build.gradle:42]: fix android build 064.4 https://github.com/facebook/react-native/issues/35210

File: src/OsAuth/OsAuthScreen/OsAuthScreen.tsx [hunk 2]
```diff
@@ -14,7 +14,6 @@ type Props = {
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   buttons: Array<any>
   onGoBack?: () => void
-  error?: null | false | string
```
Comment (stackchain) [inline src/OsAuth/OsAuthScreen/OsAuthScreen.tsx:17]: get rid of the red almost unreadable error msg on screen.

File: src/auth/authError.ts [hunk 3]
```diff
@@ -1,15 +1,34 @@
...
 
-  const decoder = (error: Error | null | undefined) => {
-    if (!error || /code: 13/.test(error?.message)) return // 13 = Cancelled by user
-    if (/code: 7/.test(error?.message)) return strings.tooManyAttempts
-    return strings.unknownError
-  }
+  return Platform.select({
+    android: (error: any): string | undefined => {
+      if (!error) return
+      if (/code: 13/.test(error?.message)) return
```
Comment (stackchain) [inline src/auth/authError.ts]: android, the errors code are inside the `.message` 13 // cancelled by the user

File: src/auth/authError.ts [hunk 4]
```diff
@@ -1,15 +1,34 @@
...
+
+      // iOS it will fallback to PIN, if wrong pin, sensor would be disabled (app will trigger pin creation)
+      if (/code: 7/.test(error?.message)) return strings.tooManyAttempts
+
+      return `${strings.unknownError}`
+    },
+
+    ios: (error: any): string | undefined => {
+      if (!error) return
+      if (error?.code === -128) return
```
Comment (stackchain) [inline src/auth/authError.ts]: ios, the error obj has `.code` property, -128 // cancelled by the user

File: src/i18n/locales/en-US.json [hunk 5]
```diff
@@ -114,7 +114,7 @@
   "components.send.biometricauthscreen.NOT_RECOGNIZED": "Biometrics were not recognized. Try again",
   "components.send.biometricauthscreen.SENSOR_LOCKOUT": "Too many failed attempts. The sensor is now disabled",
   "components.send.biometricauthscreen.SENSOR_LOCKOUT_PERMANENT": "Your biometrics sensor has been permanently locked. Use an alternate login method.",
-  "components.send.biometricauthscreen.UNKNOWN_ERROR": "Unknown error",
+  "components.send.biometricauthscreen.UNKNOWN_ERROR": "Something went wrong, try again later. Check your app settings.",
```
Comment (stackchain) [inline src/i18n/locales/en-US.json:117]: improved the message Denis' suggestion

**Review by banklesss — APPROVED** [review]

File: src/index.tsx [hunk 6]
```diff
@@ -92,22 +89,3 @@ const AppWithProviders = () => {
 }
 
 AppRegistry.registerComponent(appName, () => AppWithProviders)
-
-const useMigrations = () => {
```
Comment (stackchain) [inline src/index.tsx:96]: moved into -> Storage/migrations

File: src/i18n/LanguageProvider.tsx [hunk 7]
```diff
@@ -44,6 +45,7 @@ const useLanguageCode = ({onSuccess, ...options}: UseQueryOptions<string> = {})
       return defaultLanguageCode
     },
     onSuccess: (languageCode) => {
+      crashReporting.setStringValue('language_code', languageCode)
```
Comment (stackchain) [inline src/i18n/LanguageProvider.tsx]: migrated from actions

File: src/hooks/index.ts [hunk 8]
```diff
@@ -945,29 +892,3 @@ export const useResync = (wallet: YoroiWallet, options?: UseMutationOptions<void
     resync: mutation.mutate,
   }
 }
-export const useAuthSetting = (storage: Storage, options?: UseQueryOptions<AuthSetting, Error>) => {
```
Comment (stackchain) [inline src/hooks/index.ts:950]: moved to -> auth/hooks

File: src/Storage/StorageProvider.tsx [hunk 9]
```diff
@@ -12,3 +12,10 @@ export const useStorage = () => React.useContext(StorageContext) || invalid()
 const invalid = () => {
   throw new Error('Missing StorageProvider')
 }
+
+export const SettingsStorageKeys = {
+  OldAuthWithOs: '/appSettings/isSystemAuthEnabled', // deprecated, need for migrations
```
Comment (wolverineks) [inline src/Storage/StorageProvider.tsx]: move to migrations

File: src/AppNavigator.tsx [hunk 10]
```diff
@@ -30,20 +36,26 @@ export const AppNavigator = () => {
   const strings = useStrings()
   const storage = useStorage()
   const isMaintenance = useSelector(isMaintenanceSelector)
+  const {alert} = useAuthOsError()
```
Comment (wolverineks) [inline src/AppNavigator.tsx]: combine with useAuthWithOS

File: src/Storage/migrations/authSetting.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
+import {AUTH_WITH_OS, AUTH_WITH_PIN, disableAllEasyConfirmation, getAuthSetting} from '../../auth'
```
Comment (wolverineks) [inline src/Storage/migrations/authSetting.ts]: do something with version naming?

File: src/i18n/LanguageProvider.tsx (hunk 7 above)
Comment (wolverineks) [inline src/i18n/LanguageProvider.tsx]: matching query key for invalidations

Comment (wolverineks) [inline src/i18n/LanguageProvider.tsx]:
```suggestion
```

File: src/i18n/LanguageProvider.tsx [hunk 12]
```diff
@@ -4,6 +4,7 @@ import {IntlProvider} from 'react-intl'
 import {NativeModules, Platform, Text} from 'react-native'
 import {useMutation, UseMutationOptions, useQuery, useQueryClient, UseQueryOptions} from 'react-query'
 
+import crashReporting from '../legacy/crashReporting'
```
Comment (wolverineks) [inline src/i18n/LanguageProvider.tsx]:
```suggestion
```

File: src/i18n/LanguageProvider.tsx [hunk 13]
```diff
@@ -4,6 +4,7 @@ import {IntlProvider} from 'react-intl'
 import {NativeModules, Platform, Text} from 'react-native'
```
Comment (wolverineks) [inline src/i18n/LanguageProvider.tsx]: matching query key in invalidation

File: src/Storage/migrations/4_9_0.test.ts [hunk 14]
```diff
@@ -56,6 +59,8 @@ describe('migrateAuthSetting', () => {
 
     await migrateAuthSetting(storage)
 
-    await expect(await storage.getItem(AUTH_SETTINGS_KEY)).toBe(os)
+    await expect(await storage.getItem(SettingsStorageKeys.Auth)).toBe(os)
   })
 })
+
+const OLD_OS_AUTH_KEY = '/appSettings/isSystemAuthEnabled'
```
Comment (wolverineks) [inline src/Storage/migrations/4_9_0.test.ts]: import from migrations

File: src/Storage/migrations/4_9_0.ts [hunk 15]
```diff
@@ -18,3 +21,7 @@ export const migrateAuthSetting = async (storage: Storage) => {
     return storage.setItem(SettingsStorageKeys.Auth, JSON.stringify(AUTH_WITH_PIN))
   }
 }
+
+export const to4_9_0 = async () => migrateAuthSetting(storage)
```
Comment (wolverineks) [inline src/Storage/migrations/4_9_0.ts]: is this necessary?

File: src/Storage/migrations/4_9_0.ts [hunk 16]
```diff
@@ -18,3 +21,7 @@ export const migrateAuthSetting = async (storage: Storage) => {
...
+
+const OLD_OS_AUTH_KEY = '/appSettings/isSystemAuthEnabled'
```
Comment (wolverineks) [inline src/Storage/migrations/4_9_0.ts]:
```suggestion
export const OLD_OS_AUTH_KEY = '/appSettings/isSystemAuthEnabled'
```

File: src/Storage/migrations/4_9_0.ts (hunk 15 above)
Comment (wolverineks) [reply src/Storage/migrations/4_9_0.ts]:
```suggestion
export const to4_9_0 = migrateAuthSetting
```

File: src/Storage/migrations/hooks.ts [hunk 17]
```diff
@@ -14,7 +13,7 @@ export const useMigrations = () => {
 
       // asc order
       // 4.9.0
-      if (before4_9_0) await migrateAuthSetting(storage) // old auth settings
+      if (before4_9_0) await to4_9_0()
```
Comment (wolverineks) [inline src/Storage/migrations/hooks.ts]: why not pass in storage?

File: src/auth/hooks.ts [hunk 18]
```diff
@@ -40,33 +39,52 @@ export const useAuthOsEnabled = (options?: UseQueryOptions<boolean, Error>) => {
...
 
 export const useAuthWithOs = (
-  {authenticationPrompt}: {authenticationPrompt: AuthenticationPrompt; storage: Storage},
   options?: UseMutationOptions<void, Error>,
+  {authenticationPrompt}: {authenticationPrompt?: AuthenticationPrompt} = {},
 ) => {
+  const strings = useStrings()
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const alert = (error: any) => {
```
Comment (wolverineks) [inline src/auth/hooks.ts:69]:
```suggestion
  const alert = (error: Error) => {
```

Comment (wolverineks) [inline src/auth/hooks.ts]:
```suggestion
  const alert = (error: unknown) => {
```

File: src/auth/hooks.ts [hunk 19]
```diff
@@ -40,33 +39,52 @@ export const useAuthOsEnabled = (options?: UseQueryOptions<boolean, Error>) => {
...
+    if (error instanceof Keychain.Errors.CancelledByUser) return
+    if (error instanceof Keychain.Errors.TooManyAttempts) return Alert.alert(strings.error, strings.tooManyAttempts)
+    return Alert.alert(strings.error, strings.unknownError)
+  }
+
+  const primaryAuthenticationPrompt: AuthenticationPrompt = {
```
Comment (wolverineks) [inline src/auth/hooks.ts]:
```suggestion
  const defaultAuthenticationPrompt: AuthenticationPrompt = {
```

File: src/auth/hooks.ts [hunk 20]
```diff
@@ -40,33 +39,52 @@ export const useAuthOsEnabled = (options?: UseQueryOptions<boolean, Error>) => {
...
+    cancel: strings.cancel,
+    title: strings.authorize,
+  }
+
   const mutation = useMutationWithInvalidations({
     ...options,
     invalidateQueries: [['authSetting']],
-    mutationFn: () => Keychain.authenticate(authenticationPrompt),
+    mutationFn: () => Keychain.authenticate(authenticationPrompt ?? primaryAuthenticationPrompt),
```
Comment (wolverineks) [inline src/auth/hooks.ts]:
```suggestion
    mutationFn: () => Keychain.authenticate(authenticationPrompt ?? defaultAuthenticationPrompt),
```

File: src/Storage/migrations/4_9_0.ts (hunk 15 above)
Comment (stackchain) [reply src/Storage/migrations/4_9_0.ts]: I thinking about a interface for all migrations like `migrate` , going with the suggestion for now

File: src/auth/hooks.ts [hunk 21]
```diff
@@ -0,0 +1,319 @@
...
+  return {...mutation, enableAuthWithOs}
+}
+
+export const useAuthWithOs = (
+  options?: UseMutationOptions<void, Error>,
+  {authenticationPrompt}: {authenticationPrompt?: AuthenticationPrompt} = {},
+) => {
+  const strings = useStrings()
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (wolverineks) [inline src/auth/hooks.ts]:
```suggestion
```

File: src/auth/hooks.ts [hunk 22]
```diff
@@ -0,0 +1,319 @@
...
+}
+
+export const useAuthOsWithEasyConfirmation = (
+  {id, authenticationPrompt}: {id: YoroiWallet['id']; authenticationPrompt?: AuthenticationPrompt},
+  options?: UseMutationOptions<string, Error>,
+) => {
+  const strings = useStrings()
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const alert = (error: any) => {
```
Comment (wolverineks) [inline src/auth/hooks.ts]:
```suggestion
  const alert = (error: unknown) => {
```

File: src/auth/hooks.ts [hunk 23]
```diff
@@ -0,0 +1,319 @@
...
+  }
+}
+
+export const useAuthOsWithEasyConfirmation = (
+  {id, authenticationPrompt}: {id: YoroiWallet['id']; authenticationPrompt?: AuthenticationPrompt},
+  options?: UseMutationOptions<string, Error>,
+) => {
+  const strings = useStrings()
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (wolverineks) [inline src/auth/hooks.ts]:
```suggestion
```

File: src/auth/hooks.ts [hunk 24]
```diff
@@ -0,0 +1,319 @@
...
+  const alert = (error: any) => {
+    if (error instanceof Keychain.Errors.CancelledByUser) return
+    if (error instanceof Keychain.Errors.TooManyAttempts) return Alert.alert(strings.error, strings.tooManyAttempts)
+    return Alert.alert(strings.error, strings.unknownError)
+  }
+
+  const primaryAuthenticationPrompt: AuthenticationPrompt = {
```
Comment (wolverineks) [inline src/auth/hooks.ts]:
```suggestion
  const defaultAuthenticationPrompt: AuthenticationPrompt = {
```

File: src/auth/hooks.ts [hunk 25]
```diff
@@ -0,0 +1,319 @@
...
+    cancel: strings.cancel,
+    title: strings.authorize,
+  }
+
+  const mutation = useMutation({
+    ...options,
+    mutationFn: () => Keychain.getWalletKey(id, authenticationPrompt ?? primaryAuthenticationPrompt),
```
Comment (wolverineks) [inline src/auth/hooks.ts]:
```suggestion
    mutationFn: () => Keychain.getWalletKey(id, authenticationPrompt ?? defaultAuthenticationPrompt),
```

File: src/auth/hooks.ts [hunk 26]
```diff
@@ -0,0 +1,319 @@
...
+  }
+
+  const defaultAuthenticationPrompt: AuthenticationPrompt = {
+    cancel: strings.cancel,
+    title: strings.authorize,
+  }
+
+  const mutation = useMutationWithInvalidations({
+    ...options,
+    invalidateQueries: [['authSetting']],
```
Comment (wolverineks) [inline src/auth/hooks.ts]: why does this need to invalidate?

Comment (stackchain) [reply src/auth/hooks.ts]: it should be moved into `useEnableAuthWithOs`

Comment (stackchain) [reply src/auth/hooks.ts]: part as success process

Comment (stackchain) [reply src/auth/hooks.ts]: https://github.com/Emurgo/yoroi-mobile/pull/2213/commits/e77da4fc7c883dad571b864fc5eb8d68278af5a4

File: src/auth/hooks.ts [hunk 27]
```diff
@@ -0,0 +1,319 @@
...
+    },
+  })
+
+  return {
+    ...mutation,
+    authWithOs: mutation.mutate,
+  }
+}
+
+export const useDisableEasyConfirmation = (wallet: YoroiWallet, options?: UseMutationOptions) => {
```
Comment (wolverineks) [inline src/auth/hooks.ts:128]: i think disabling easy confirmation should require authWithOS

File: src/auth/hooks.ts [hunk 28]
```diff
@@ -0,0 +1,319 @@
...
+    invalidateQueries: [['walletMetas']],
+  })
+
+  return {
+    ...mutation,
+    disableEasyConfirmation: mutation.mutate,
+  }
+}
+
+export const useEnableEasyConfirmation = (wallet: YoroiWallet, options?: UseMutationOptions<void, Error, string>) => {
```
Comment (wolverineks) [inline src/auth/hooks.ts:141]: should enabling easy confirmation require auth with os?

File: src/auth/hooks.ts (hunk 27 above)
Comment (stackchain) [reply src/auth/hooks.ts:128]: easy confirmation is per wallet, (it reads the wallet - key), authWithOS, is per App

File: src/auth/hooks.ts (hunk 26 above)
Comment (wolverineks) [inline src/auth/hooks.ts]:
```suggestion
```

**Pull request merged**